File: Data Structures & Algorithms/sqrtx/submission-4.java
class Solution {
    public int mySqrt(int x) {
        int left = 1;
        int right = x;

        int root = 0;
        while (left <= right) {
            long mid = left + (right - left) / 2;

            if (mid * mid == x) {
                return (int) mid;
            } else if (mid * mid < x) {
                root = (int) mid;
                left = (int) mid + 1;
            } else {
                right = (int) mid - 1;
            }
        }

        return root;
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-4.kt
class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        nums.sort()
        val triplets = mutableListOf<List<Int>>()

        for (pivot in 0..nums.size - 3) {
            if (pivot == 0 || nums[pivot] != nums[pivot - 1]) {
                var left = pivot + 1
                var right = nums.size - 1

                while (left < right) {
                    val sum = nums[left] + nums[right] + nums[pivot]
                    if (sum == 0) {
                        triplets.add(listOf(nums[pivot], nums[left], nums[right]))

                        do {
                            left++
                        } while (left < right && nums[left] == nums[left - 1])
                        do {
                            right--
                        } while (left < right && nums[right] == nums[right + 1])
                    } else if (sum < 0) {
                        left++
                    } else {
                        right--
                    }
                }
            }
        }

        return triplets
    }
}
